src/v/serde/json/detail/tests/string_test.cc

```diff
@@ -0,0 +1,224 @@
+  // Valid string with trailer.
+  {"\"aaa\"aaa", string_parser::result::done, 5, "aaa"},
+  // Valid string with escaped control character.
+  {R"("a\ta")", string_parser::result::done, 6, "a\ta"},
+  // Invalid string because of unescaped control character.
+  {"\"a\ta\"", string_parser::result::invalid_json_string, 3},
+
+  // Below are some of the more interesting strings from jsonchecker test
+  // suite.
+  {"\"1~!@#$%^&*()_+-={':[,]}|;.</>?\"",
+   string_parser::result::done,
+   32,
+   "1~!@#$%^&*()_+-={':[,]}|;.</>?"},
+  {R"("{\"object with 1 member\":[\"array with 1 element\"]}")",
+   string_parser::result::done,
+   55,
+   R"({"object with 1 member":["array with 1 element"]})"},
+  {
+    R"("\u0123\u4567\u89AB\uCDEF\uabcd\uef4A")",
+    string_parser::result::done,
+    38,
+    "ģ䕧覫췯ꯍ",
+  },
+  {
+    R"("\/\\\"\uCAFE\uBABE\uAB98\uFCDE\ubcda\uef4A\b\f\n\r\t`1~!@#$%^&*()_+-=[]{}|;:',./<>?")",
+    string_parser::result::done,
+    85,
+    "/\\\"\xEC\xAB\xBE\xEB\xAA\xBE\xEA\xAE\x98\xEF\xB3\x9E\xEB\xB3\x9A\xEE"
+    "\xBD"
+    "\x8A\b\f\n\r\t`1~!@#$%^&*()_+-=[]{}|;:',./<>?",
+  },
+
+  ////////////////////////////////////////////////////////////
+  // Unicode escape sequences from RapidJSON test suite.
+
+  // G clef sign U+1D11E (𝄞)
+  {
+    R"("\uD834\uDD1E")",
+    string_parser::result::done,
+    14,
+    "\xF0\x9D\x84\x9E",
+  },
+
+  // Single low surrogate pair in string is invalid.
+  {
+    R"("\udc4d")",
+    string_parser::result::invalid_json_string,
+    7,
+  },
+
+  // Invalid surrogate pair.
+  {
+    R"("\uD800X")",
+    string_parser::result::invalid_json_string,
+    8,
+  },
+  {
+    R"("\uD800\uFFFF")",
+    string_parser::result::invalid_json_string,
+    13,
+  },
+
+  //////////////////////////////////////////////////////////////
+  // Custom test cases.
+
+  // Short unicode escape sequence.
+  {
+    R"("\uCAF")",
+    string_parser::result::invalid_json_string,
+    7,
+    "",
+  },
+});
+
+TEST_P(string_parse_test, test_string) {
+    auto [input, expected_err, expected_pos, expected_output] = GetParam();
+    string_parser parser;
+    string_parser::result err;
+    ss::temporary_buffer<char> buf(input.data(), input.size());
+    size_t pos = parser.advance(buf, err);
+    EXPECT_EQ(pos, expected_pos);
+    EXPECT_EQ(err, expected_err);
+    if (expected_err == string_parser::result::done) {
+        EXPECT_EQ(iobuf_as_string(std::move(parser).value()), expected_output);
+    } else {
+        EXPECT_THROW(std::move(parser).value(), std::runtime_error);
+    }
+}
+
+TEST_P(string_parse_test, test_piecewise) {
+    auto [input, expected_err, expected_pos, expected_output] = GetParam();
+
+    if (expected_err != string_parser::result::done) {
+        // Skip cases that end in error because we don't know how to split them
+        // in interesting but practical ways.
+        return;
+    }
+
+    // Try splitting the input at every possible position.
```

Comment:
Love it. Was going to suggest testing the power set of split points (all combos of valid splits) but this is probably just as effective and cheaper